src/crimson/os/seastore/cache.cc

```diff
@@ -1820,11 +1820,32 @@ Cache::replay_delta(
       DEBUG("replay extent delta at {} {} ... -- {}, prv_extent={}",
             journal_seq, record_base, delta, *extent);
 
-      assert(extent->last_committed_crc == delta.prev_crc);
-      assert(extent->version == delta.pversion);
-      extent->apply_delta_and_adjust_crc(record_base, delta.bl);
-      extent->set_modify_time(modify_time);
-      assert(extent->last_committed_crc == delta.final_crc);
+      if (delta.paddr.get_addr_type() == paddr_types_t::SEGMENT ||
+	  get_extent_category(delta.type) != data_category_t::DATA) {
+	assert(extent->last_committed_crc == delta.prev_crc);
+	assert(extent->version == delta.pversion);
+	extent->apply_delta_and_adjust_crc(record_base, delta.bl);
+	extent->set_modify_time(modify_time);
+	assert(extent->last_committed_crc == delta.final_crc);
+      } else {
+	assert(delta.paddr.get_addr_type() == paddr_types_t::RANDOM_BLOCK);
+	extent->apply_delta_and_adjust_crc(record_base, delta.bl);
+	extent->set_modify_time(modify_time);
+	if (delta.pversion == 0 && extent->version > 0 &&
+	  extent->last_committed_crc == delta.final_crc) {
+	  // if an extent was rewritten in an inplace manner---
+	  // note that the extent after being rewritten was a clean state,
+	  // pversion can be 0 and exent->version > 0 because 
+	  // clean extent's version after read_extent() is 0
+	  if (extent->is_dirty()) {
+	    remove_from_dirty(extent);
+	    extent->state = CachedExtent::extent_state_t::CLEAN;
+	  }
+	  extent->version = 0;
+	} else if (delta.final_crc == extent->last_committed_crc) {
+	  assert(extent->version == delta.pversion);
+	}
+      }
```

Comment:
I think the following lines are still necessary:

```
if (delta.final_crc == extent->last_committed_crc) {
 assert(extent->version == delta.pversion);
}
```

because TRIM_DIRTY can conflict with other transactions after the extent is rewritten in an inplace manner.
This cause a crc mismatch because the extent has been already updated.

Reply:
> This cause a crc mismatch because the extent has been already updated.

I see. I think that may cause intermediate crc mismatches, but the crc should finally be matched after replaying all the deltas. So for inplace-rewrite extents, the final crc needs to be verified at the end of the replay, right?

Reply:
Right. So, we need something like a map that has < CachedExtentRef, final CRC > to compare the CRC at the end of the replay, right?

Reply:
I think so, let's introduce a temporary map only for the extents that can do inplace-rewrite, and skip calculating their intermediate CRCs.

Reply:
Can you take a look at the last commit? I tried to address your comment.